docs/metrics.md

```diff
@@ -0,0 +1,14 @@
+# Build Controller Metrics
+
+The Build component exposes several metrics to help you monitor the health and behavior of your build resources.
+
+Following build metrics are exposed at service `build-operator-metrics` on port `8383`.
+
+| Name | Type | Description | Status |
+| ---- | ---- | ----------- | ------ |
+| `up_status` | Gauge | The up status of the build-controller. | experimental |
+| `build_count` | Gauge | Build creation count. | experimental |
+| `buildrun_count` | Gauge | BuildRun execution count. | experimental |
+| `buildrun_establish_duration` | Gauge | BuildRun establish duration in seconds. | experimental |
+| `buildrun_ready_duration` | Gauge | BuildRun ready duration in seconds. | experimental |
+| `buildrun_ready_duration_his_[start, width, count]` | Histogram | BuildRun ready duration histogram in seconds. | experimental |
```

Comment:
Why the histogram and gauge about this? What's the difference?

The naming best practice about this would be just the `_duration_seconds` suffix. And for me reading this, I don't actually understand what this metric tells me.

Reply:
- buildrun_ready_duration: I used it for the trend to show how long recent build runs take
- buildrun_ready_duration_his: I used it for the statistics to classify different execution times to different buckets

<img width="566" alt="截屏2020-06-10 下午1 59 23" src="https://user-images.githubusercontent.com/18108042/84232325-a42ef500-ab22-11ea-8471-81d3e3806b16.png">

Reply:
These sound like they should be instrumented in the component that actually performs the build run.

Reply:
What is your mean? They are all calculate during buildrun controller execution logic

Reply:
The controller doesn't actually execute the build. The component that does should expose that metric.

Reply:
Yep, I also compared other community metric design, like tekton pipeline:
https://github.com/tektoncd/pipeline/blob/master/docs/metrics.md

The duration metrics are all `Histogram` types, so I also removed the `Gauge` one, just keep the `Histogram` one:
```
	buildRunEstablishDuration = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name: "build_buildrun_establish_duration_seconds",
			Help: "BuildRun establish duration in seconds.",
			Buckets: prometheus.LinearBuckets(0, 0.5, 10),
		},
		[]string{"build_buildstrategy", "buildrun_namespace"})

	buildRunCompletionDuration = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "build_buildrun_completion_duration_seconds",
			Help:    "BuildRun completion duration in seconds.",
			Buckets: prometheus.LinearBuckets(50, 50, 10),
		},
		[]string{"build_buildstrategy", "buildrun_namespace"})
```

The change is in the latest commit: https://github.com/redhat-developer/build/pull/235/commits/248b3db1c98c9c36a87cab7f6f75f2095cb8400a

Reply:
I'm a little confused. If tekton already has these metrics, why do we need them again?

Reply:
No, Tekton just has the metrics about `taskrun`, they don't have our metrics like buildrun, I just compare the metric types with Tekton metrics.

Reply:
I see, I didn't realize they are distinct concepts.